File: ExternalControllerFramework/Matlab/extCtrl.m
clear; clc;

% Simulation & UDP setup
destIP   = "127.0.0.1";
destPort = 65000;
localPort = 65009;

udpSender = dsp.UDPSender('RemoteIPAddress', destIP, ...
                           'RemoteIPPort', destPort, ...,
                           'LocalIPPortSource', 'Property', ...
                           'LocalIPPort', localPort);

udpReceiver = dsp.UDPReceiver('LocalIPPort', localPort, ...
                              'RemoteIPAddress', destIP, ...
                              'MessageDataType', 'double', ...
                              'MaximumMessageLength', 12);

% Control variables
ctrlTime        = 0.0;
dtCtrl          = 0.05;
sim_idx         = 0;
prev_sim_idx    = 0;
send_ctrl = true;
no_meas_count = 0;

% Initial Control Command
wellNames = { ...
    'BM01 Land', ...
    'BM02 Deepwater MPD', ...
    'BM03 Deepwater CML'};
[selectedWell, ok] = listdlg( ...
    'PromptString','Select test well:', ...
    'SelectionMode','single', ...
    'ListString',wellNames, ...
    'ListSize',[300 100]);
if ~ok
    error('No test well selected.');
end
TestWellMenu = selectedWell;
disp(['Selected well: ', wellNames{TestWellMenu}]);

switch TestWellMenu
    case 1
        OP.z_c0  = [0.4356; 0];
    case 2
        OP.z_c0  = [0.5871; 0];
    case 3
        OP.z_c0  = [0.5871; 0];
end

z_uA = OP.z_c0(1);
z_uB = OP.z_c0(2);
w_uA = 0; 
w_uB = 0;
E_ctrl = 0;

disp('Controller running... Ctrl+C to stop');
while true
    % 1) SEND CONTROL COMMAND
    if send_ctrl
        controlCmd = [sim_idx; z_uA; z_uB; w_uA; w_uB; E_ctrl];  % 6 doubles
        udpSender(controlCmd);
        send_ctrl = false;
    end
    
    % 2) RECEIVE MEASUREMENTS
    measurements = udpReceiver();
    if isempty(measurements)
        no_meas_count = no_meas_count +1;
        pause(0.001);
        if (no_meas_count >= 5000)
            sim_idx = 0;
            prev_sim_idx = 0;
            z_uA = OP.z_c0(1);
            z_uB = OP.z_c0(2);
            w_uA = 0; 
            w_uB = 0;
            E_ctrl = 0;
            send_ctrl = true;
            no_meas_count = 0;
        end
        continue;
    end
    t_sim    = measurements(1);
    sim_idx   = measurements(2);
    p_c_r     = measurements(3);
    p_c       = measurements(4);
    p_stp     = measurements(5);
    q_p       = measurements(6);
    q_bl      = measurements(7);
    q_c       = measurements(8);
    z_cA      = measurements(9);
    z_cB      = measurements(10);
    w_cA      = measurements(11);
    w_cB      = measurements(12);


    % 3) CONTROL LAW
    if sim_idx > prev_sim_idx
        K_p = -0.001;
        w_uA = K_p * (p_c_r - p_c);

        fprintf("\n[MEASUREMENT] t=%.4f, count=%d, p_c=%.4f, p_c_r=%.4f, p_stp=%.4f, q_p=%.4f, q_bl=%.4f, q_c=%.4f, z_cA=%.4f, z_cB=%.4f, w_cA=%.4f, w_cB=%.4f", ...
                t_sim, sim_idx, p_c, p_c_r, p_stp, q_p, q_bl, q_c, z_cA, z_cB, w_cA, w_cB);

        ctrlTime = ctrlTime + dtCtrl;
        prev_sim_idx = sim_idx;
        send_ctrl = true;
    end


end

File: ExternalControllerFramework/Python/extCtrl.py
import socket
import struct
import time
import sys

LOCAL_IP = '127.0.0.1'
LOCAL_PORT = 65008
REMOTE_IP = '127.0.0.1'
REMOTE_PORT = 65000 
TIMEOUT_COUNT = 5000 

COMMAND_FORMAT = '<6d' 
MEASUREMENT_FORMAT = '<12d'
MEASUREMENT_BYTES = struct.calcsize(MEASUREMENT_FORMAT)

def select_well_and_get_op():
    well_names = {
        1: 'BM01 Land',
        2: 'BM02 Deepwater MPD',
        3: 'BM03 Deepwater CML'
    }
    
    print("\n--- Select Test Well ---")
    for key, name in well_names.items():
        print(f"[{key}] {name}")
    print("------------------------")
    
    while True:
        try:
            selection = int(input("Enter number for test well: "))
            if selection in well_names:
                break
            else:
                print("Invalid selection. Please enter 1, 2, or 3.")
        except ValueError:
            print("Invalid input. Please enter a number.")

    print(f"Selected well: {well_names[selection]}")
    
    if selection == 1:
        return 0.4356, 0.0
    elif selection == 2 or selection == 3:
        return 0.5871, 0.0
    
    return 0.4356, 0.0

def run_udp_controller():
    z_cA_init, z_cB_init = select_well_and_get_op()
    
    sim_idx = 0
    prev_sim_idx = 0

    w_uA = 0.0
    w_uB = 0.0
    z_cA = z_cA_init
    z_cB = z_cB_init
    E_ctrl = 0.0
    
    send_ctrl = True
    no_meas_count = 0

    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.bind((LOCAL_IP, LOCAL_PORT))
        sock.setblocking(False) 
        
        print(f"\nController running... Ctrl+C to stop")
        print(f"Listening on port {LOCAL_PORT} and sending to {REMOTE_IP}:{REMOTE_PORT}")

    except Exception as e:
        print(f"Failed to initialize socket: {e}")
        sys.exit(1)

    while True:
        try:
            # 1) SEND CONTROL COMMAND
            if send_ctrl:
                control_cmd_values = [
                    float(sim_idx), 
                    z_cA, 
                    z_cB, 
                    w_uA,
                    w_uB,
                    E_ctrl
                ]
                
                bytes_out = struct.pack(COMMAND_FORMAT, *control_cmd_values)
                sock.sendto(bytes_out, (REMOTE_IP, REMOTE_PORT))
                send_ctrl = False
                no_meas_count = 0 
            
            # 2) RECEIVE MEASUREMENTS
            measurements = None
            try:
                data, sender = sock.recvfrom(MEASUREMENT_BYTES)
                if len(data) == MEASUREMENT_BYTES:
                    unpacked_measurements = struct.unpack(MEASUREMENT_FORMAT, data)
                    measurements = list(unpacked_measurements)
                    no_meas_count = 0
                else:
                    pass
                    
            except BlockingIOError:
                pass 
            except Exception as e:
                print(f"Socket error during receive: {e}")

            if measurements is None:
                no_meas_count += 1
                time.sleep(0.001)
                
                if no_meas_count >= TIMEOUT_COUNT:
                    print(f"Timeout (>{TIMEOUT_COUNT * 0.001}s) reached. Resetting control variables.")
                    sim_idx = 0
                    prev_sim_idx = 0
                    w_uA = 0.0
                    w_uB = 0.0
                    E_ctrl = 0.0
                    z_cA = z_cA_init
                    z_cB = z_cB_init
                    send_ctrl = True
                    no_meas_count = 0
                continue
            
            # Measurements:
            # 1:t_sim, 2:sim_idx, 3:p_c_r, 4:p_c, 5:p_stp, 6:q_p, 7:q_bl, 8:q_c, 9:z_cA, 10:z_cB, 11:w_cA, 12:w_cB
            t_sim    = measurements[0]
            sim_idx   = int(measurements[1])
            p_c_r     = measurements[2]
            p_c       = measurements[3]
            p_stp     = measurements[4]
            q_p       = measurements[5]
            q_bl      = measurements[6]
            q_c       = measurements[7]
            z_cA      = measurements[8]
            z_cB      = measurements[9]
            w_cA      = measurements[10] # Measurement A
            w_cB      = measurements[11] # Measurement B
            
            # 3) CONTROL LAW
            if sim_idx > prev_sim_idx:
                K_p = -0.001
                w_uA = K_p * (p_c_r - p_c)

                print(
                    f"\n[MEASUREMENT] t={t_sim:.4f}, count={sim_idx}, "
                    f"p_c={p_c:.4f}, p_c_r={p_c_r:.4f}, p_stp={p_stp:.4f}, "
                    f"q_p={q_p:.4f}, q_bl={q_bl:.4f}, q_c={q_c:.4f}, "
                    f"z_cA={z_cA:.4f}, z_cB={z_cB:.4f}, "
                    f"w_cA={w_cA:.4f}, w_cB={w_cB:.4f}"
                )

                prev_sim_idx = sim_idx
                send_ctrl = True
            
        except KeyboardInterrupt:
            print("\nShutting down controller.")
            sock.close()
            break
        except Exception as e:
            print(f"An unhandled error occurred in the main loop: {e}")
            sock.close()
            break

if __name__ == "__main__":
    run_udp_controller()